src/Simulation/qdk_sim_rs/docs/c-api.md

```diff
@@ -0,0 +1,147 @@
+# Using Experimental Simulators from C
+
+This module exposes a C API for this crate, useful for embedding into simulation
+runtimes.
+
+## Safety
+
+As this is a foreign-function interface, many of the functions exposed here
+are **unsafe**, representing that the caller is required to ensure that safety
+conditions in the host language are upheld.
+
+Please pay attention to any listed safety notes when calling into this C
+API.
+
+## Generating and Using C API Headers
+
+The [`qdk_sim`](..) crate has enabled the use of [`cbindgen`](https://crates.io/crates/cbindgen), such that C-language header files are generated automatically as part of the build for this crate.
+
+```bash
+cargo install --force cbindgen
+cbindgen --language C --output include/qdk_sim.h
+cbindgen --language C++ --output include/qdk_sim.hpp
+```
+
+This will generate `include/qdk_sim.h` and `include/qdk_sim.hpp, which can then be used from C and C++ callers, respectively. For example, to call from C:
+
+```c
+#include <stdio.h>
+#include "qdk_sim.h"
+
+int main() {
+    uintptr_t sim_id;
+    uintptr_t result0, result1;
+
+    init(2, "mixed", &sim_id);
+    h(sim_id, 0);
+    h(sim_id, 1);
+
+    m(sim_id, 0, &result0);
+    m(sim_id, 1, &result1);
+
+    printf("got %llu %llu", result0, result1);
+    destroy(sim_id);
+}
+```
```

Comment:
I may not know the full picture, and my note can be not applicable, but just in case I'll mention that:
* `uintptr_t` is an [optional type](https://www.cplusplus.com/reference/cstdint/).
* `uintptr_t` has architecture-dependent size, same as pointers (32-bit on 32-bit architecture, 64-bit on 64-bit architecture)

I would not recommend to use `uintptr_t` across the ABI boundary.

Reply:
In this case, the example uses the types in `qdk_sim.h`, as generated by cbindgen. E.g., the autogenerated `qdk_sim.h` contains the following declaration:

```C
#include <stdarg.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>

// ...

int64_t init(uintptr_t initial_capacity, const char *representation, uintptr_t *sim_id_out);
```

My suggestion as to how to best fix that would be to change `usize` declarations in `c_api.rs` to `u64`. Since that could affect our e2e builds, I'd prefer to not make that change this close to release but to defer to improvements in the feature post-release.

Reply:
Whatever you choose to do is totally fine with me. I just want to make sure that you are fully aware of the risks if I see any.

Reply:
Makes sense! I'll add that note to the same issue for tracking C API improvements along with deallocation handling and granular return codes.

Reply:
https://github.com/microsoft/qsharp-runtime/issues/732